Source/Classes/PINURLSessionManager.m

```diff
@@ -187,6 +193,60 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didComp
     });
 }
 
+- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didFinishCollectingMetrics:(NSURLSessionTaskMetrics *)metrics
+{
+    NSDate *requestStart = [NSDate distantFuture];
+    NSDate *firstByte = [NSDate distantPast];
+    
+    BOOL valid = YES;
+    
+    for (NSURLSessionTaskTransactionMetrics *metric in metrics.transactionMetrics) {
+        if (metric.requestStartDate == nil || metric.responseStartDate == nil) {
+            //Only evaluate requests which completed their first byte.
+            valid = NO;
+            break;
+        }
+        if ([requestStart compare:metric.requestStartDate] != NSOrderedAscending) {
+            requestStart = metric.requestStartDate;
+        }
+        if ([firstByte compare:metric.responseStartDate] != NSOrderedDescending) {
+            firstByte = metric.responseStartDate;
+        }
+    }
+    
+    if (valid) {
+        [self storeTimeToFirstByte:[firstByte timeIntervalSinceDate:requestStart] forHost:task.originalRequest.URL.host];
+    }
+}
+
+- (void)storeTimeToFirstByte:(NSTimeInterval)timeToFirstByte forHost:(NSString *)host
+{
+    [self lock];
```

Comment:
I wonder if you can get away without locking. `NSCache` supports concurrent access, and the logic below is sufficiently fuzzy that missing/overwriting a measurement probably won't be super meaningful.